### ISSUE TITLE
Git Add-Patch Interactive Mode

### ISSUE THREAD

**Issue by TheBrenny** [gitkraken/vscode-gitlens#1679]
Sometimes I work on multiple bugs or features without changing branches or because that wastes time. Instead, I do all my work at once, and then when something's ready to commit and push, I quickly flick through my changed files to find all the changes that relate to each other. If the file has multiple unrelated changes I go to my terminal and run `git add -p ./file/to/commit`. Sometimes my file has a number of changes to it becomes tedious to flick through each change.

Similar to how Git Lens has an interactive rebase editor, I'm proposing an interactive Add-Patch editor. I imagine something similar to a diff view would pop up, and we could click on code lenses for each hunk to either add it to the changes, or split it into smaller hunks. From there, the right hand editor would be a R/W editor allowing us to edit the hunks (like how the `e` option of `git add -p` lets you edit hunks).

I've duplicated this issue on the [VSCode Repo](https://github.com/microsoft/vscode/issues/134728) because I think that it _could_ be something that's supported out-of-the-box, but considering it's also an enhanced and niche type of feature, I've made this feature request as well.

**Comment by ridvanaltun** [comment]
Ping

I think is the most needed feature, is there any progress?

**Comment by eamodio** [comment]
I would love to add a nice hunks view for staging in a more interactive way -- so this is a great feature suggestion.

At the same time, you can do this today using the built-in Git support in a couple of ways:

With the changed file open you can click on the SCM Glyph in the "gutter", and you will get a peek view showing the change, and you can choose to stage or revert.

![image](https://user-images.githubusercontent.com/641685/163106301-966d5dad-c56d-496a-8abf-9ecd6e65a9d2.png)

Or, if you open a diff of a changed file, you can use the diff editor, and either select lines or place your cursor on a line and run the `Stage Selected Ranges` command, either in the menu shown below or from the command palette or create a custom shortcut key for it like I have (that's why it shows with `alt+s`)

![image](https://user-images.githubusercontent.com/641685/163106432-8fa4452e-1380-42b1-a3d9-8522af107734.png)

**Comment by TheBrenny** [comment]
Yup, fair enough, I'll close! That solution literally solves the problem, and solves it to the T of my description. 🎉

**Comment by eamodio** [comment]
I'm going to keep it open as I do want to provide a more discoverable and strictly hunk based view for this. Ideally even across files.

**Comment by Hacker1337** [comment]
It would be great to have an integrated in diff view tool to add patch or reset with checkbox, like in pycharm IDE, 
![image](https://github.com/gitkraken/vscode-gitlens/assets/40338902/66e9c3a3-6b1a-4588-b82e-fba57d3c55df)
(from pycharm [doc](https://www.jetbrains.com/help/pycharm/viewing-changes-information.html#comparing_local_changes))
It has super convinient tools for git. Looking forward to see them implemented in vscode.
